### ISSUE TITLE
certbot-auto initial install: some transparency, please?

### ISSUE THREAD

**Issue by IzzySoft** [certbot/certbot#3731]
We use certs a.o. to enhance privacy (making it harder to "snoop" our connections). But to ensure *our* privacy, it would be good to have some more transparency:

We're expected to run `certbot-auto` as `root`, and it then "will take care for the rest automatically". While this is nice and relaxing from the users' workload, I'd like to know in advance what gets installed where to. Some examples:

* which directories will be created, and which files placed where?
* will it install its "own stuff" just into the directory it's invoked from?
* would it "mess up" things when run from `/usr/local/bin` (i.e. by installing a bunch of dependencies there)?
* it looks like it always wants to install all its plugins, no matter what (e.g. I see no way to tell it I don't want/need the Apache plugin, so it doesn't need to install augeas). Is that so? The docs say nothing about it (or I've missed that).

Of course one could try to "reverse-engineer" those details from the `certbot-auto` script. But that's a) not feasible (considering many users do not have the technical background to do so) and b) should not be needed with an open-source program which is all about privacy and security.

**TL;DR:** Can we have at least a raw listing of what will be installed where to (raw e.g. in telling "Python modules go into `/some/dir`, without listing all their files) – and a way to tell which plugins we (don't) want?

This shouldn't be a rant, just reasoning :) Wondering noone brought it up before (or did I miss that as well?)

Thanks in advance!

**Comment by pde** [comment]
* This [FAQ entry](https://certbot.eff.org/faq/#does-certbot-require-root-administrator-privileges) answers the question about root and paths.
* For most use cases (`--csr` is a special historical exception), certbot likes to output files in a structured way to `/etc/letsencrypt` because that allows much better housekeeping for tasks like renewal. You can move that with `--config-dir`.

`certbot-auto` is a hacked-together package manager that aims to work in a secure and automated way on as many platforms as possible. If you want more transparency and control, you can install a copy of Certbot from most OSes now, or you can use `pip` to install the Pythonic version (though we'd recommend making a Python virtual environment to do that in a way that won't leave specific versions of dependencies lying around in `/usr/local/lib/` that can confuse your host OS python packages).

**Comment by IzzySoft** [comment]
Thanks @pde – but sorry, no, that does *not* answer the question about root and paths (just about those required at runtime, but not those for the initial install). Reading through the script I e.g. see it

* wants to write to `/etc/apt/sources.list`
* using apt-get to install OS packages
* writes several `.py` scripts and runs them
* obviously installs several Python packages (assumingly in its own virtualenv, but I'm not sure that's for all packages or if some are installed globally)

Sure I could analyze all that myself by e.g. running it in a clean VM, checking the before and after states – if I had the time :) I don't need to count every single file/byte, but would like a raw idea of what to expect *before* running the script. Rawly, something like:

> on install, `certbot-auto` …

> * uses the OS's package manager to install {these packages}. For this, it might need to update your systems `/etc/apt/sources.list` / {corresponding RPM etc. files}
> * does write files to the following locations {list-of-directories}
> * most things will be kept in a Python virtualenv located at `$XDG_DATA_HOME/letsencrypt` (default: `~/.local/share/letsencrypt`)
> * ?? will be placed in your `$PATH` (`/usr/local/bin` / ??)

> You can run `certbot-auto` initially from anywhere you like; unless it's in one of the locations mentioned, no files will be placed there (so you can e.g. place it right in your `$PATH`, e.g. `/usr/local/bin`, for convenience). Altogether, depending on your system and which requirements are already met, installation will need up to about X MB additional space.

Which locations is written to while using the installed certbot is documented fine.

Please don't get me wrong. But with package managers one can easily inspect the packages or have the pm list what's installed where. Also an uninstall, if needed, is relatively easy to achieve. If I had to move certbot to a different machine, I'd have no idea at all what to cleanup on the "old" machine, and only a rough idea of where to look for *most* of the things. Right now, I could only tell about the locations used at runtime, plus the `$XDG_DATA_HOME/letsencrypt` directory – but I cannot tell what packages (OS & Python) might be obsolete if I no longer use certbot on the "old" machine. I prefer keeping machines clean: software not installed cannot pose security risks, and does not need to be updated :)

**Comment by IzzySoft** [comment]
Well, so I answer it myself. On install,

* certbot-auto doesn't install anything into the directory it's called from (unless you call it from a directory it would write to anyway).
* On Debian/derivates, it installs a couple of packages: `augeas-lenses libaugeas0 libc-dev-bin libc6-dev libexpat1-dev libffi-dev libpython-dev libpython2.7-dev libssl-dev python-dev python-pip-whl python-virtualenv python2.7-dev python3-virtualenv virtualenv zlib1g-dev`, resulting in about 70 MB required diskspace
* It checks for (and installs, if not present) another couple of packages. These are usually already present in most systems (many of them being "base requirements" anyway): `ca-certificates gcc python dialog`. If those were all not present (very unlikely), that would result in another ~200 MB. Considering noone runs a server without the certificates (and their openssl dependency), take away at least half of that.

That's about what I was after. Not sure if I missed something because it "ran silently", so updates to my notes are welcome.
